After revert: Defer local updates if there are missing updates and only call `GetMissingOnyxMessages` once

chrispader opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

@danieldoglas 

### Details
<!-- Explanation of the change or anything fishy that is going on -->

This PR applies back the changes from https://github.com/Expensify/App/pull/38997 after the revert in https://github.com/Expensify/App/pull/39668

This PR adds unit and E2E tests covering the `OnyxUpdateManager` and the new update deferral logic.

### Fixed Issues
<!---
1. Please postfix `$` with a URL link to the GitHub issue this Pull Request is fixing. For example, `$ https://github.com/Expensify/App/issues/<issueID>`.
2. Please postfix  `PROPOSAL:` with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).  For example, `PROPOSAL: https://github.com/Expensify/App/issues/<issueID>#issuecomment-1369752925`

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking and its automation will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<issueID>
$ https://github.com/Expensify/App/issues/<issueID(comment)>

Do NOT only link the issue number like this: $ #<issueID>
--->
$ https://github.com/Expensify/App/issues/38748
PROPOSAL:


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

- [x] Verify that no errors appear in the JS console

#### Test 1

I created a [testing branch](https://github.com/margelo/expensify-app-fork/tree/%40chrispader/GetMissingOnyxMessages-deferring-testing) that prevents some updates from being applied (therefore `GetMissingOnyxUpdates` will be triggered) and has some useful logs.

Either test in this branch or on your own. In [`OnyxUpdateManager`](https://github.com/Expensify/App/blob/main/src/libs/actions/OnyxUpdateManager.ts) do the following:

1. Delay the call to `applyDeferredUpdates` with a `setTimeout`
2. Prevent some update from being applied
3. Check that `GetMissingOnyxUpdates` is triggered
4. Add some more updates
5. Check that these updates are added to the `deferredUpdates` object
6. Check that no more calls to `GetMissingOnyxUpdates` are performed
7. Once the timeout has ended, check that both the missing and deferred updates are applied correctly.


#### Test 2
1. Create a chat report with a peer
2. Create a policy room between both of you
3. after those rooms are created, and both of you opened them at least once, comment these calls to make sure we're not reloading the data from the backend instead: https://github.com/Expensify/App/blob/a4d01a87b242eef06af37db965bd135745e342e5/src/libs/actions/Report.ts#L769-L789
4. Now send messages in both chat rooms, alternating between them, and fast, in a way that you can notice if anything is missing (so both start sending some kind of ordered messages, like sequential numbers)
5. Make sure that none of the messages were lost in those chats.

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->

Not needed.

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

- [ ] Verify that no errors appear in the JS console
- Open a chat between two testers
- Both testers should send consecutive messages at the same time (e.g. 1, 2, 3, 4) 
- No message should be lost between them
- No more than 1 `GetMissingOnyxMessages` request should be executed at the same time

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<!-- add screenshots or videos here -->

<img width="557" alt="Screenshot 2024-03-27 at 16 05 13" src="https://github.com/Expensify/App/assets/20173411/56863211-c81f-454e-aa4c-5a24cf298aba">


</details>

<details>
<summary>Android: mWeb Chrome</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: Native</summary>

<!-- add screenshots or videos here -->

![Simulator Screenshot - iPhone 15 Pro - 2024-03-27 at 16 10 34](https://github.com/Expensify/App/assets/20173411/a5580be6-c9f9-4ffe-a845-7f1741261474)

</details>

<details>
<summary>iOS: mWeb Safari</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

<!-- add screenshots or videos here -->

<img width="1164" alt="Screenshot 2024-03-27 at 16 13 55" src="https://github.com/Expensify/App/assets/20173411/a744d1fa-1e0c-4604-b7fe-d8eeb8620a52">

</details>

<details>
<summary>MacOS: Desktop</summary>

<!-- add screenshots or videos here -->

</details>

Tested deferring logic on web, iOS and Android. mWeb behavior is the same as in web...

Commit:
Revert "Merge pull request #39668 from Expensify/revert-38997-@chrispader/prevent-simultaneous-calls-to-GetMissingOnyxMessages"

This reverts commit 3864cdbbcbd672d5d434076c5ebf79c63f1bf4d7, reversing
changes made to 42ee04cdeb957c2234ef7f0aad5499fd805f5f3f.

Commit:
remove queue clearing and add TS type guard for OnyxUpdate

Commit:
fix: type check

Commit:
simplify code

Commit:
simplify code

Commit:
fix: and simplify

Commit:
add comment

Commit:
rename variables

hungvu193 commented:
Take a look on test build 👀

eh2077 commented:
@chrispader Could you kindly elaborate how this PR fixes those deploy blocker issues? Thanks!

hungvu193 commented:
BUG: It always redirects to NotFound page after refreshing a page.

https://github.com/Expensify/App/assets/16502320/d047b828-1814-4357-9fbe-e03715b1497b

chrispader commented:
> @chrispader Could you kindly elaborate how this PR fixes those deploy blocker issues? Thanks!

I cannot say if the change fixes the problem, because this seems to only happen on very-high traffic accounts.

I removed [this check](https://github.com/Expensify/App/blob/8fe465d72fb95f50e9df8888dafaae55fff8364f/src/libs/actions/OnyxUpdateManager.ts#L39-L51) for invalid update formats, which unpauses the `SequentialQueue` which might cause the problem

danieldoglas commented:
@hungvu193 I think that's an issue unrelated to this PR, it happens on the PR testing domain

danieldoglas commented:
@eh2077 @hungvu193 You'll need to try to simulate a situation with a lot of data going on between you. sending lots of messages on groups both of you are and etc.  It's gonna be a tough one to test.

hungvu193 commented:
@danieldoglas  Can you create a slack thread to discuss about testing this PR? It would be better

chrispader commented:
> @danieldoglas Can you create a slack thread to discuss about testing this PR? It would be better

i asked about that here: https://expensify.slack.com/archives/C049HHMV9SM/p1712330961862649?thread_ts=1712264151.436019&cid=C049HHMV9SM

@danieldoglas could you check if there are (other) people that have time and can test this with their accounts?

eh2077 commented:
I tried to reproduce this DB issue https://github.com/Expensify/App/issues/39650 but failed to dupe it. I use high traffic accounts.

eh2077 commented:
> I tried to reproduce this DB issue #39650 but failed to dupe it. I use high traffic accounts.

@chrispader I managed to reproduce #39650 using two high traffic accounts. It's not stable but I reproduced it by following steps
Preq: Account A on mobile device, Account B on Chrome Web
1. Send several messages from B to A
2. Send several messages from A to B
3. Observed that messages are received normally
4. **_Wait a few minutes_**
5. Send several messages from A to B
6. Observed that, from B, the last message is shown in LHN but is not shown in the chat
<img width="1241" alt="image" src="https://github.com/Expensify/App/assets/117511920/f601d656-6553-485f-bd02-86a685228a1c">

chrispader commented:
> > I tried to reproduce this DB issue #39650 but failed to dupe it. I use high traffic accounts.
> 
> @chrispader I managed to reproduce #39650 using two high traffic accounts. It's not stable but I reproduced it by following steps Preq: Account A on mobile device, Account B on Chrome Web
> 
> 1. Send several messages from B to A
> 2. Send several messages from A to B
> 3. Observed that messages are received normally
> 4. **_Wait a few minutes_**
> 5. Send several messages from A to B
> 6. Observed that, from B, the last message is shown in LHN but is not shown in the chat

you reproduced this with the current branch still?

danieldoglas commented:
@chrispader any movement on this one? What's the ETA?

Commit:
Merge branch 'main' into @chrispader/GetMissingOnyxMessages-deferred-updates-after-revert

Commit:
remove unnecessary pause

Commit:
check for isLoadingApp and ActiveClientManager first

Commit:
Merge branch 'main' into @chrispader/GetMissingOnyxMessages-deferred-updates-after-revert

chrispader commented:
> > I tried to reproduce this DB issue #39650 but failed to dupe it. I use high traffic accounts.
> 
> @chrispader I managed to reproduce #39650 using two high traffic accounts. It's not stable but I reproduced it by following steps Preq: Account A on mobile device, Account B on Chrome Web
> 
> 1. Send several messages from B to A
> 2. Send several messages from A to B
> 3. Observed that messages are received normally
> 4. **_Wait a few minutes_**
> 5. Send several messages from A to B
> 6. Observed that, from B, the last message is shown in LHN but is not shown in the chat

Unfortunately i cannot reproduce this. I see that when sending a lot of messages (from both accounts) the chat flickers and refreshes a lot of times, but in the end i always have the latest updates.

I tested this with two high-traffic accounts on both

- 1. device web, 2. device web
- 1. device Android, 2. device web

eh2077 commented:
@chrispader Thanks for the update. I'll review and re-test it and post an update tmr.

roryabraham reviewed (commented):
high-level feedback is that this code looks like it could really benefit from some automated test coverage

eh2077 commented:
@chrispader I pulled latest changes of this branch and I caught a different inconsistency - LHN message is not updated. I'm not sure if it's related to this change or it's a different bug.

<img width="1401" alt="Screenshot 2024-04-16 at 10 34 17 AM" src="https://github.com/Expensify/App/assets/117511920/b045c0b4-dfb3-4f3d-a547-9ac764b50f9c">

Commit:
Merge branch 'main' into @chrispader/GetMissingOnyxMessages-deferred-updates-after-revert

chrispader commented:
unfortunately i cannot reproduce any of these issues with 2 high-traffic accounts 🥲

also, i feel like the problems shown in both recordings stem from a different reason, this seems to be related to logic that handles the "New messages" toast and divider line that you can click. Once you clicked it, the content in LHN got updated.

![Screenshot 2024-04-16 at 12 17 41](https://github.com/Expensify/App/assets/20173411/4e36d041-0866-4e47-b6e0-6f3385e30f20)

chrispader commented:
> high-level feedback is that this code looks like it could really benefit from some automated test coverage

yes that's true. @danieldoglas and i were planning on creating a follow-up PR which adds tests and more documentation (maybe a flow diagram) that explains this logic even more.

roryabraham commented:
I hear that a lot when I ask for tests on complex PRs: "yes we absolutely should add tests, and that's why we'll do it after", and then about 30% of the time that actually happens.

My two cents: just add the tests. What you lose in time now you'll gain in not having to redo this, and making it easy to add coverage for any edge cases you missed here.

danieldoglas commented:
I think this PR was delayed for a long time already, so we can wait one more week on it while adding tests. let's do that @chrispader

chrispader commented:
> I think this PR was delayed for a long time already, so we can wait one more week on it while adding tests. let's do that @chrispader 

i'm gonna work on the tests tmrw!

Commit:
Merge branch 'main' into @chrispader/GetMissingOnyxMessages-deferred-updates-after-revert

Commit:
minor improvements of OnyxUpdateMangaer

Commit:
implement tests

Commit:
add exports for testing

Commit:
fix: test implementation

Commit:
implement proxy for OnyxUpdateManager

Commit:
update tests

Commit:
extract util logic from OnyxUpdateManager

Commit:
extract createProxyForValue and further split up utils

Commit:
fix: tests

Commit:
extract mocks from test

Commit:
apply updaes

Commit:
remove unnecessary promise trigger code

Commit:
update tests and mocks

Commit:
Merge branch 'main' into @chrispader/GetMissingOnyxMessages-deferred-updates-after-revert

chrispader commented:
I now added tests for the `OnyxUpdateManager`. These test the deferral logic inside `OnyxUpdateManager` with mocked Onyx update application. Is this sufficient, or do we need to also test this E2E with actual Onyx data?

cc @danieldoglas @quinthar

Commit:
fix: ts

eh2077 commented:
@chrispader I found that tests is failed. Can you take a look?
<img width="452" alt="image" src="https://github.com/Expensify/App/assets/117511920/4d9f9b62-f06c-4a5a-8a4b-95200e65f987">

Commit:
fix: tests

Commit:
move specific mocks to mock folders

Commit:
fix: update mocks and tests

Commit:
simplify

chrispader commented:
i'm gonna add a few more "E2E" tests, where we actually apply data to Onyx and run the whole flow starting from `applyOnyxUpdatesReliably`

Commit:
fix: update mocks, tests and test utils to allow for broader E2E tests

Commit:
add more tests and comments

Commit:
remove waitForBatchedChanges

Commit:
rename and move helper function

Commit:
add isPaused checker function to SequentialQueue

Commit:
reset more things before each test

Commit:
add another test

Commit:
fix: first test

chrispader commented:
@danieldoglas @quinthar @arosiclair i've now also added some broader "end-to-end" test under `tests/actions/OnyxUpdateManagerTests`, that cover all the logic from calling `applyOnyxUpdatesRealiably` to the updates being written to storage with Onyx.

These tests also ensure, that the `SequentialQueue` is paused and the updates have been applied in the correct order.

There are also unit tests in `tests/unit/OnyxUpdateManagerTets` that test only the logic in `OnyxUpdateManager` without references/dependencies to other components in E/App. These tests are heavily mocked, which means there are no network calls and onyx writes, whereas we also check for correct values in the `deferredUpdates` array.

I think this PR should now be ready for a final round of reviews and hopefully merge. Lmk if you think there are any more useful tests or missed edge-cases.

hungvu193 commented:
@danieldoglas Can you please apply `Ready to Build` label again?

danieldoglas commented:
Triggering build

hungvu193 commented:
@chrispader Can I still use the test branch you mentioned in the test steps? I saw it's a little bit outdated.

Commit:
Merge branch 'main' into @chrispader/GetMissingOnyxMessages-deferred-updates-after-revert

Commit:
remove unused line

danieldoglas commented:
bump @hungvu193 @eh2077

hungvu193 commented:
I took few tests today, no problem so far, but I still need to take another few tests ti make sure it works 

https://github.com/Expensify/App/assets/16502320/72d80f7e-a7c8-4a6e-9447-39b287f6bbb9

eh2077 commented:
Thanks for the ping. It's EOD for me and I'll review and test it early tmr.

hungvu193 commented:
**Test 2** worked perfectly for me however I found issue with **Test 1** where we received Pusher event but the reaction didn't update, please see the video below, mostly at the end of the video.


https://github.com/Expensify/App/assets/16502320/a22fc31e-8f72-4411-a08b-cbd6954c50ab

hungvu193 commented:
Precondition: Checkout[ testing branch](https://github.com/margelo/expensify-app-fork/tree/%40chrispader/GetMissingOnyxMessages-deferring-testing)
Steps to reproduce:
1. Create a policy room between 2 users.
2. User1 send some messages then added few edits and add reactions
3. Observe update from User2,

Commit:
Merge branch 'main' into @chrispader/GetMissingOnyxMessages-deferred-updates-after-revert

chrispader commented:
> Precondition: Checkout[ testing branch](https://github.com/margelo/expensify-app-fork/tree/%40chrispader/GetMissingOnyxMessages-deferring-testing) Steps to reproduce:
> 
> 1. Create a policy room between 2 users.
> 2. User1 send some messages then added few edits and add reactions
> 3. Observe update from User2,

hmm, again unable to reproduce unfortunately :/

i did exactly as described in your steps and as seen in your video. I sent a bunch of messages from one user to the other, then in the other direction. Updated messages and added reactions. After a while (because a in the testing branch updates a delayed and some are dropped on purpose) the app is able to fetch all missing updates with `GetMissingOnyxUpdates`.

Do you have any more specific repro steps? Still curious if this is actually caused by this PR and if i need some special sort of account, to repro this.

<img width="2632" alt="Screenshot 2024-04-26 at 23 34 39" src="https://github.com/Expensify/App/assets/20173411/124a8591-c582-49dc-b129-797e272d6d4d">

chrispader commented:
@hungvu193 i'm also not 100% sure if this isn't somehow caused by my changes in the testing branch by the way i intentionally drop updates and delay them.

chrispader commented:
JFYI @danieldoglas @hungvu193 @arosiclair 
I'm gonna be OOO from 28/04 until 7-8/04. If we cannot resolve this issue until then, i'm gonna have to hand the PR over to one of my colleagues from Margelo.

hungvu193 commented:
> @hungvu193 i'm also not 100% sure if this isn't somehow caused by my changes in the testing branch by the way i intentionally drop updates and delay them.

Can you try these steps?

1. Send message from 1 to 9.
2. Edit messages 9 to 5.
3. Wait for all updates to be applied.
4. Add and remove few reactions
5. Observer the result

chrispader commented:
> > @hungvu193 i'm also not 100% sure if this isn't somehow caused by my changes in the testing branch by the way i intentionally drop updates and delay them.
> 
> Can you try these steps?
> 
> 1. Send message from 1 to 9.
> 2. Edit messages 9 to 5.
> 3. Wait for all updates to be applied.
> 4. Add and remove few reactions
> 5. Observer the result
> 
> I saw the timeout, however I waited for few minutes and still didn't see the updated. I'll take another test later today

Ahh i see now what might be the problem in the testing branch. I basically drop every third and fourth update, to simulate missing updates. If one or both of these (omitted) updates are at the end of all the updates, they will never be applied, because there is no later update that triggeres the `GetMissingOnyxMessages` flow.

hungvu193 commented:
> In the testing branch in `applyOnyxUpdatesReliably` if you set the `shouldOmitUpdate` flag to `false`, this issue should not happen anymore. Can you confirm that?

you mean `shouldRunSync` to false right?
https://github.com/Expensify/App/blob/10fc359d9fdd1e05e03a4769e2a73f95e703be51/src/libs/actions/applyOnyxUpdatesReliably.ts#L14

Commit:
Merge branch 'main' into @chrispader/GetMissingOnyxMessages-deferred-updates-after-revert

danieldoglas commented:
Nice. Seems like this is almost there... @hungvu193 @eh2077 you think we're good to merge this?

hungvu193 commented:
> Nice. Seems like this is almost there... @hungvu193 @eh2077 you think we're good to merge this?

Sounds good to me 😄

eh2077 commented on src/libs/actions/OnyxUpdateManager/utils/index.ts:
```diff
@@ -0,0 +1,137 @@
+    // re-fetch the missing updates and then apply the remaining deferred updates after the gap
+    if (latestMissingUpdateID) {
+        return new Promise((resolve, reject) => {
+            deferredUpdatesProxy.deferredUpdates = {};
```
I think we should move this line to right after calling the synchronous method `detectGapsAndSplit`.

eh2077 replied:
Just note for myself
> The executor is called synchronously (as soon as the Promise is constructed) with the resolveFunc and rejectFunc functions as arguments.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/Promise

We also early return when existingDeferredUpdatesCount > 0

So there won't be an issue that updates are discarded without being applied.

eh2077 commented on src/libs/actions/OnyxUpdateManager/utils/index.ts:
```diff
@@ -0,0 +1,137 @@
+        return Promise.resolve();
+    }
+
+    const {applicableUpdates, updatesAfterGaps, latestMissingUpdateID} = detectGapsAndSplit(pendingDeferredUpdates, lastUpdateIDFromClient);
```
Should we reset `deferredUpdatesProxy.deferredUpdates` after this line?

eh2077 commented on src/libs/actions/OnyxUpdateManager/index.ts:
```diff
@@ -38,17 +39,40 @@ Onyx.connect({
+
+const resetDeferralLogicVariables = () => {
+    queryPromise = undefined;
+    deferredUpdatesProxy.deferredUpdates = {};
```
I think we don't need to reset here. We only need to reset it in method `validateAndApplyDeferredUpdates`

eh2077 reviewed (commented):
@danieldoglas Sorry for the late update. 

The PR tested well while I have concerns about resetting `deferredUpdatesProxy.deferredUpdates`. 

If only resetting `deferredUpdatesProxy.deferredUpdates` after calling method `detectGapsAndSplit`, the two tests are also passed.
```
npm test -- tests/unit/OnyxUpdateManagerTest.ts
npm test -- tests/actions/OnyxUpdateManagerTest.ts
```

I'd like to hear your thoughts on this.

chrispader commented:
> If only resetting `deferredUpdatesProxy.deferredUpdates` after calling method `detectGapsAndSplit`, the two tests are also passed.

i'm not sure i understand what you mean. what scneria would resetting deferred updates after `detectGapsAnsSplit` simulate?

The deferred updates should/can only be reset by the `OnyxUpdateManager`.

Do you mean we should protect this export, so manipulation from the outside is not possible?

eh2077 commented:
@chrispader thanks for your comment. Sorry, please omit my comments as I overlooked and misunderstood something.

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/8904499086) to staging by https://github.com/danieldoglas in version: 1.4.69-0 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

kbecciv commented:
@chrispader @danieldoglas QA team is blocked to verify this PR, having the following error when trying to run the code snippet in console. Can you please verify internally?
![Screenshot 2024-05-01 at 9 23 17 at night](https://github.com/Expensify/App/assets/93399543/64204086-553f-4423-bcdd-c278a86ca237)

francoisl commented:
The QA steps shouldn't involve commenting code out, can you guys think of an alternative way to QA this internally please?

danieldoglas commented:
@kbecciv Updated the tests

kbecciv commented:
@danieldoglas 

This PR is failing because of issue https://github.com/Expensify/App/issues/38748 - there are multiple GetMissingOnyxMessages requests in request tab.

The issue is checked in: Web and Desktop

![image](https://github.com/Expensify/App/assets/93399543/89f7d1b6-f5c5-4a79-a9db-93c4a7e36cb7)

https://github.com/Expensify/App/assets/93399543/8a3963ef-7d7f-4713-8ee0-f5d1fade9dd0

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/8924174545) to production by https://github.com/Beamanator in version: 1.4.69-2 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

danieldoglas commented:
That actually seems correct - it did 3, but one at a time, not 3 at the same time.

chrispader commented:
@danieldoglas i didn't follow recent conversations 100%, but i think the implementation in this PR is still valid and there's no need to fix anything right now, right?

danieldoglas commented:
Yep, this is all correct.
